src/queue.rs

```diff
@@ -131,111 +148,38 @@ pub struct DescriptorChain<M: GuestAddressSpace> {
     mem: M::T,
     desc_table: GuestAddress,
     queue_size: u16,
-    ttl: u16, // used to prevent infinite chain cycles
-
-    /// The current descriptor
-    desc: Descriptor,
-    curr_indirect: Option<Box<DescriptorChain<M>>>,
-    is_master: bool,
+    next_index: u16,
+    ttl: u16,
+    is_indirect: bool,
 }
 
 impl<M: GuestAddressSpace> DescriptorChain<M> {
-    fn read_new(
+    fn with_ttl(
         mem: M::T,
         desc_table: GuestAddress,
         queue_size: u16,
         ttl: u16,
         index: u16,
-    ) -> Option<Self> {
-        if index >= queue_size {
-            return None;
-        }
-
-        let desc_size = size_of::<Descriptor>();
-        let desc_addr = desc_table.checked_add(desc_size as u64 * index as u64)?;
-        let desc = mem.read_obj(desc_addr).ok()?;
-        let chain = DescriptorChain {
+    ) -> Self {
+        DescriptorChain {
             mem,
             desc_table,
             queue_size,
+            next_index: index,
             ttl,
-            desc,
-            curr_indirect: None,
-            is_master: true,
-        };
-
-        if chain.is_valid() {
-            Some(chain)
-        } else {
-            None
+            is_indirect: false,
         }
     }
 
-    /// Create a new DescriptorChain instance.
-    fn checked_new(
-        mem: M::T,
-        dtable_addr: GuestAddress,
-        queue_size: u16,
-        index: u16,
-    ) -> Option<Self> {
-        Self::read_new(mem, dtable_addr, queue_size, queue_size, index)
+    /// Create a new `DescriptorChain` instance.
+    fn new(mem: M::T, desc_table: GuestAddress, queue_size: u16, index: u16) -> Self {
```

Comment:
Slightly unrelated, but it'd be nice to have a public function for building a `DescriptorChain`, so other crates can make use of it for their own unit tests.

Reply:
I think we should keep auxiliary logic separate from the "main" functionality, so the latter is smaller and thus easier to review, test, and audit. We can add methods for building descriptor chains and other useful things in a separate module (made opt-in via a feature) that can be then consumed by our tests and potentially other crates as well. For example, we used this approach in [event-manager](https://github.com/rust-vmm/event-manager) where the `test_utilities` feature gates the `src/utilities` module. How does that sound?

Reply:
Just wanted to call out that the terrible thing about the test_utilities in event-manager is that `cargo test` no longer works. You need to do either `cargo test --all-features` or `cargo test --features test_utilities`.

Otherwise I agree with what Alex said :))

Reply:
Makes sense to me. It's cleaner and I don't see any significant drawbacks. It most situations you already need to pass the `--all-features` flag anyways.